proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+checking if a trait is enabled, the allowed characters for a trait name have to
+be restricted. The following rules are enforced on trait names:
+
+- The first character must be a [Unicode XID start
+  character](https://unicode.org/reports/tr31/#Figure_Code_Point_Categories_for_Identifier_Parsing)
+  (most letters), a digit, or `_`.
+- Subsequent characters must be a [Unicode XID continue
+  character](https://unicode.org/reports/tr31/#Figure_Code_Point_Categories_for_Identifier_Parsing)
+  (a digit, `_`, or most letters), `-`, or `+`.
+
+## Impact on existing packages
+
+There is no impact on existing packages. Any package can start adopting package
+traits but in doing so **must not** move existing API behind new traits.
+
+## Future directions
+
+### Consider traits during dependency resolution
+
+The implementation to this proposal only considers traits **after** the
+dependency resolution when constructing the module graph. This is inline with
+how platform specific dependencies are currently handled. In the future, both
+platform specific dependencies and traits can be taken into consideration during
+dependency resolution.
+
+### Integrated compiler trait checking
+
+The current proposal passes enabled traits via custom defines to the compiler
+and code can check it using regular define checks (`#if DEFINE`). In the future,
+we can extend the compiler to make it aware of package traits to allows syntax
+like `#if trait(FOO)` or implement an extensible configuration macro similar to
+Rust's `cfg` macro.
+
+## Alternatives considered
+
+### Different naming
+
+During the implementation and writing of the proposal different names for
+_package traits_ have been considered such as _package features_. A lot of the
+other considered names have other meanings in the language already.
+
+## Prior art
+
+Other dependency managers have similar features to control optional dependencies
+and conditional compilation.
+
+- [Cargo](https://doc.rust-lang.org/cargo/) has [optional features](https://doc.rust-lang.org/cargo/reference/features.html) that allow conditional compilation and optional dependencies.
+- [Maven](https://maven.apache.org/) has [optional dependencies](https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html).
+- [Gradle](https://gradle.org/) has [feature variants](https://docs.gradle.org/current/userguide/feature_variants.html) that allow conditional compilation and optional dependencies.
+- [Go](https://golang.org/) has [build constraints](https://golang.org/pkg/go/build/#hdr-Build_Constraints) which can conditionally include a file.
+- [pip](https://pypi.org/project/pip/) dependencies can have [constraints](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-dependencies) and can have [extras](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies).
```

Comment:
The two setuptools links just go to `setuptools.html` without any scrolling for me. I also don't see `constraints` or `extras` mentioned on that page from a find in page.